pkg/pool-stable-phantom/contracts/StablePoolRates.sol

```diff
@@ -0,0 +1,319 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import "@balancer-labs/v2-interfaces/contracts/solidity-utils/openzeppelin/IERC20.sol";
+import "@balancer-labs/v2-interfaces/contracts/pool-utils/IRateProvider.sol";
+
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/InputHelpers.sol";
+import "@balancer-labs/v2-pool-utils/contracts/rates/PriceRateCache.sol";
+
+import "./StablePoolStorage.sol";
+
+abstract contract StablePoolRates is StablePoolStorage {
+    using PriceRateCache for bytes32;
+    using FixedPoint for uint256;
+
+    struct RatesParams {
+        IERC20[] tokens;
+        IRateProvider[] rateProviders;
+        uint256[] tokenRateCacheDurations;
+    }
+
+    // This contract uses timestamps to slowly update its Amplification parameter over time. These changes must occur
+    // over a minimum time period much larger than the blocktime, making timestamp manipulation a non-issue.
+    // solhint-disable not-rely-on-time
+
+    // Token rate caches are used to avoid querying the price rate for a token every time we need to work with it.
+    // The "old rate" field is used for precise protocol fee calculation, to ensure that token yield is only
+    // "taxed" once. The data structure is as follows:
+    //
+    // [ expires | duration | old rate | current rate ]
+    // [ uint32  |  uint32  |  uint96  |   uint96     ]
+
+    mapping(IERC20 => bytes32) internal _tokenRateCaches;
+
+    event TokenRateCacheUpdated(IERC20 indexed token, uint256 rate);
+    event TokenRateProviderSet(IERC20 indexed token, IRateProvider indexed provider, uint256 cacheDuration);
+
+    constructor(StablePoolStorage.StorageParams memory storageParams, RatesParams memory rateParams)
+        StablePoolStorage(storageParams)
+    {
```

Comment:
Let's not call `StablePoolStorage`'s constructor here, and instead leave it up to the parent contract (like we do in e.g. `EOASignaturesValidator`, which doesn't initialize `EIP712`).